Add pg ilike search to posts title and body

todtb opened the pull request:
#3540

- Add pg ilike search to `posts.title` and `posts.body`
- Add search form to `/journal` (browse)
- Add post search results page which shows results, paginated

## Browse:
![image](https://user-images.githubusercontent.com/19367605/200447385-046670c6-39b1-48cf-9bb7-fef070e04ec1.png)

## Results:
![image](https://user-images.githubusercontent.com/19367605/200447468-010a99ba-29eb-4051-b518-f820ecde2cc9.png)

![image](https://user-images.githubusercontent.com/19367605/200447504-5562af37-b725-4bd9-b3ae-de3f389e3dcb.png)

kueda commented on config/routes.rb:
```diff
@@ -499,6 +499,7 @@
       get :for_user
     end
   end
+  get "/journal_search", to: "posts#search"
```
Let's make this `/posts/search` instead.

kueda reviewed (changes requested):
Many thanks, working for me, and sort of meets my original need, but not quite. To do that I'd like the following:

- [ ] Add the search boxes to posts/index (right-aligned in posts/_header would work for me)
- [ ] posts/search should take `parent_type` and `parent_id` params to scope search to a particular journal (doesn't have to be implemented that way, but I want journal-specific search so I can find particular posts in the iNat (site) journal, or my (user) journal)
- [ ] posts/search should show what journal you're searching

Sorry, another case of me writing an under-specified issue.

todtb commented:
Ok, reworked this a bit so that it applies to all `parent_types` (`User`, `Project`, `Site`). Since I was doing this I took the opportunity to create a `HasJournal` concern for those models, thus helping clean up the controller flow a bit.

Also removed original "Browse" search as I see that wasn't really the primary goal here and searching ALL might not be that helpful.

N.B.
- Does change `Project`
`has_many :journal_posts, class_name: "Post", as: :parent`
to
`has_many :journal_posts, class_name: "Post", as: :parent, dependent: :destroy`
Though I'm assuming it should be this way and consistent with other models now, I highlight his as it is a change.

## Project (new)
![image](https://user-images.githubusercontent.com/19367605/200978856-e3ec6cd2-3eca-4706-b7cc-c3c3af36d691.png)
## Project (old)
![image](https://user-images.githubusercontent.com/19367605/200978896-ed498d47-e0af-489b-b446-fe044ca6a9d3.png)
## Site
![image](https://user-images.githubusercontent.com/19367605/200978940-b17ef723-5252-4563-8253-2f14a6ef90ab.png)
## User (self)
![image](https://user-images.githubusercontent.com/19367605/200979046-e8703471-b406-4280-a874-88ef5347d387.png)
## User (other)
![image](https://user-images.githubusercontent.com/19367605/200979309-9485509f-4fbc-4e89-aee0-3d0fbb175d2e.png)

kueda commented on app/views/posts/search.html.haml:
```diff
@@ -0,0 +1,21 @@
+- content_for :title do
+  = @title = "#{t(:find_posts_for)} #{@parent.journal_display_name}"
```
Translators will not like this b/c they can't control the position of the journal name, they won't be able to control for journals that begin with a vowel or consonant, and they won't be able to control for gender of the journal name. Here's an alternative that skirts two of these issues:

```ruby
= @title = t( :find_posts_for, vow_or_con: @parent.journal_display_name[0].downcase, journal_name: @parent.journal_display_name )
```

And then in en.yml:

```yaml
en:
  find_posts_for: "Find Posts for: %{journal_name}"
```

Gender could still be a problem, but there's no perfect way to deal with that with arbitrary journal names.

todtb replied:
Excellent call, and a great lesson for thinking in i18n

kueda commented on app/views/posts/_header.html.haml:
```diff
@@ -2,32 +2,59 @@
+          - if @display_project.curated_by?(current_user)
+            = link_to t(:new_post), new_project_journal_post_path(@parent.journal_slug), id: "newpostbutton",
+                class: "default button"
+        = render partial: 'search_form', locals: { parent_type: @parent.class.name, parent_id: @parent.id }
```
I'm not a fan of the extraneous whitespace this creates below the title when the viewer can't see the "New Post" button. Could we remove the search form from the header at put it at the top of the right column instead?

todtb replied:
With the right column being the "archives" section? Something like:
## own
![image](https://user-images.githubusercontent.com/19367605/201531428-aaa08361-1375-468c-bc2d-67cc06537419.png)

## other user
![image](https://user-images.githubusercontent.com/19367605/201531441-5401cd07-d54b-4089-8357-e7b483ff40d9.png)

## project
![image](https://user-images.githubusercontent.com/19367605/201531591-e7df5182-f091-4a1e-86d8-6086fba74608.png)


## w/o archives:
![image](https://user-images.githubusercontent.com/19367605/201531487-3dd077a0-d1a1-4598-a446-5a188c806198.png)
### blog
![image](https://user-images.githubusercontent.com/19367605/201531554-51c028a4-d086-4390-a79a-c174c9624208.png)

kueda replied:
Exactly. Those screenshots look good to me.

kueda reviewed (changes requested):
Works great, and I like the concern. Just a few more changes and I think we'll be there.

kueda commented:
Thanks again!
